diff -ur offlineimap-6.3.3.orig/docs/Makefile offlineimap-6.3.3/docs/Makefile
--- offlineimap-6.3.3.orig/docs/Makefile	2011-03-17 19:00:55.737262335 +0100
+++ offlineimap-6.3.3/docs/Makefile	2011-03-17 19:01:44.675082102 +0100
@@ -5,8 +5,8 @@
 HTML_TARGETS = $(patsubst %.rst,%.html,$(SOURCES))
 
 RM = rm
-RST2HTML=`type rst2html >/dev/null 2>&1 && echo rst2html || echo rst2html.py`
-RST2MAN=`type rst2man >/dev/null 2>&1 && echo rst2man || echo rst2man.py`
+RST2HTML=rst2html
+RST2MAN=rst2man
 
 all: html
 
diff -ur offlineimap-6.3.3.orig/Makefile offlineimap-6.3.3/Makefile
--- offlineimap-6.3.3.orig/Makefile	2011-03-17 19:00:55.737262335 +0100
+++ offlineimap-6.3.3/Makefile	2011-03-17 19:01:22.907710668 +0100
@@ -18,7 +18,7 @@
 VERSION=`./offlineimap.py --version`
 TARGZ=offlineimap_$(VERSION).tar.gz
 SHELL=/bin/bash
-RST2HTML=`type rst2html >/dev/null 2>&1 && echo rst2html || echo rst2html.py`
+RST2HTML=rst2html
 
 all: build
 
